Reduce log size

pvollan opened the pull request:
#### a83b4253d3e9b4a9355268f68b8c9055a1b08d6a
<pre>
Reduce log size
<a href="https://bugs.webkit.org/show_bug.cgi?id=285385">https://bugs.webkit.org/show_bug.cgi?id=285385</a>
<a href="https://rdar.apple.com/142358119">rdar://142358119</a>

Reviewed by Chris Dumez.

Reduce log size by migrating more HTML Media logging to efficient version.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::play):
(WebCore::HTMLMediaElement::playInternal):
* Source/WebCore/platform/LogMessages.in:

Canonical link: <a href="https://commits.webkit.org/293577@main">https://commits.webkit.org/293577@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a33f26cf5b11de990ee2542f2fb9a7094f8101d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99290 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18938 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9187 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104420 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49889 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101331 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19227 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27371 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75565 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32667 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102297 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14608 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89642 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55925 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14402 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7623 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49250 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84327 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7710 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106778 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26403 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19249 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84525 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26766 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85844 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84037 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28700 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6386 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20130 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16156 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26343 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31534 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26164 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29477 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27731 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Scripts/generate-derived-log-sources.py:
```diff
@@ -93,11 +93,11 @@ def generate_message_receiver_implementations_file(log_messages, log_messages_re
                 file.write("    auto osLogPointer = OS_LOG_DEFAULT;\n")
             else:
-                file.write("    auto osLog = adoptOSObject(os_log_create(\"com.apple.WebKit.WebContent\", \"" + category + "\"));\n")
+                file.write("    String subsystem = makeString(\"com.apple.WebKit.WebContent.\"_s, String::number(m_pid));\n")
```
This is unnecessarily inefficient, you don't need `String::number()`, just include `<wtf/text/StringConcatenateNumbers.h>`

cdumez commented on Source/WebKit/Scripts/generate-derived-log-sources.py:
```diff
@@ -93,11 +93,11 @@ def generate_message_receiver_implementations_file(log_messages, log_messages_re
             else:
-                file.write("    auto osLog = adoptOSObject(os_log_create(\"com.apple.WebKit.WebContent\", \"" + category + "\"));\n")
+                file.write("    String subsystem = makeString(\"com.apple.WebKit.WebContent.\"_s, String::number(m_pid));\n")
+                file.write("    auto osLog = adoptOSObject(os_log_create(subsystem.utf8().data(), \"" + category + "\"));\n")
```
I think we can use `ascii()` here.

cdumez commented on Source/WebKit/Shared/LogStream.cpp:
```diff
@@ -68,13 +68,10 @@ void LogStream::logOnBehalfOfWebContent(std::span<const uint8_t> logSubsystem, s
-        auto category = byteCast<char>(logCategory.data());
-        osLog = adoptOSObject(os_log_create(subsystem, category));
-    }
+    const bool subsystemIsNullTerminated = isNullTerminated(logSubsystem);
```
no need for `const`

cdumez commented on Source/WebKit/Shared/LogStream.cpp:
```diff
@@ -68,13 +68,10 @@ void LogStream::logOnBehalfOfWebContent(std::span<const uint8_t> logSubsystem, s
-        osLog = adoptOSObject(os_log_create(subsystem, category));
-    }
+    const bool subsystemIsNullTerminated = isNullTerminated(logSubsystem);
+    String subsystem = makeString(subsystemIsNullTerminated ? logSubsystem.data() : nullptr, subsystemIsNullTerminated ? "."_s : nullptr, "WebContent."_s, String::number(m_pid));
```
Similarly, `String::number()` is not needed.

cdumez commented on Source/WebKit/Shared/LogStream.cpp:
```diff
@@ -68,13 +68,10 @@ void LogStream::logOnBehalfOfWebContent(std::span<const uint8_t> logSubsystem, s
+    const bool subsystemIsNullTerminated = isNullTerminated(logSubsystem);
+    String subsystem = makeString(subsystemIsNullTerminated ? logSubsystem.data() : nullptr, subsystemIsNullTerminated ? "."_s : nullptr, "WebContent."_s, String::number(m_pid));
+    auto category = byteCast<char>(isNullTerminated(logCategory) ? logCategory.data() : nullptr);
+    auto osLog = adoptOSObject(os_log_create(subsystem.utf8().data(), category));
```
`ascii()`

bnham commented on Source/WebKit/Shared/LogStream.cpp:
```diff
@@ -68,13 +69,10 @@ void LogStream::logOnBehalfOfWebContent(std::span<const uint8_t> logSubsystem, s
-        osLog = adoptOSObject(os_log_create(subsystem, category));
-    }
+    bool subsystemIsNullTerminated = isNullTerminated(logSubsystem);
+    String subsystem = makeString(subsystemIsNullTerminated ? logSubsystem.data() : nullptr, subsystemIsNullTerminated ? "."_s : nullptr, "WebContent."_s, m_pid);
```
This will break `WTFSignpostHandleIndirectLog`, since it is doing a pointer comparison on the subsystem/category pair which no longer matches the expected subsystem.

Let me see if I can come up with a small patch to fix it.

bnham replied:
Never mind, I'll just fix this in a follow-up patch.

pvollan replied:
Ah, great catch! I am putting this patch on hold while investigating other potential issues.

Thanks for reviewing!

cdumez commented on Source/WebKit/Scripts/generate-derived-log-sources.py:
```diff
@@ -96,9 +96,17 @@ def generate_message_receiver_implementations_file(log_messages, log_messages_re
+
             arguments_string = log_declarations_module.get_arguments_string(parameters, log_declarations_module.PARAMETER_LIST_INCLUDE_NAME | log_declarations_module.PARAMETER_LIST_MODIFY_CSTRING)
+
+            file.write("    if (m_pid < std::numeric_limits<uint16_t>::max() && m_pid >= 0)\n")
```
How much win are we really seeing from this? I'm not sure how I feel about this. It feels unfortunate to increase the complexity and encoding the PID as different types depending on its value. It also forces us to check the value every time. This is not something we usually do in our IPC.

I can't say I'm a fan.

pvollan replied:
That makes sense! I have not measured the win here, so I am not sure how much we're saving.

I can reduce the scope of the patch by just including the Media player logging changes.

Thanks for reviewing!

pvollan committed:
Reduce log size
https://bugs.webkit.org/show_bug.cgi?id=285385
rdar://142358119

Reviewed by Chris Dumez.

Reduce log size by migrating more HTML Media logging to efficient version.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::play):
(WebCore::HTMLMediaElement::playInternal):
* Source/WebCore/platform/LogMessages.in:

Canonical link: https://commits.webkit.org/293577@main
